apps/remote_control/lib/lexical/remote_control/search/indexer/source/reducer.ex

```diff
@@ -0,0 +1,145 @@
+defmodule Lexical.RemoteControl.Search.Indexer.Source.Reducer do
+  @moduledoc """
+  A module and struct that can reduce over elixir AST via Macro.prewalk/3
+
+  The reducer keeps track of blocks and parent / child relationships so extractors don't have to concern themselves
+  with the AST's overall structure, and can focus on extracting content from it.
+  """
+  alias Lexical.Document
+  alias Lexical.RemoteControl.Search.Indexer.Entry
+  alias Lexical.RemoteControl.Search.Indexer.Extractors
+  alias Lexical.RemoteControl.Search.Indexer.Metadata
+  alias Lexical.RemoteControl.Search.Indexer.Source.Block
+
+  defstruct [:entries, :document, :quoted_document, :position, :ends_at, :blocks]
+
+  @extractors [Extractors.Module]
+
+  def new(%Document{} = document, quoted_document) do
+    %__MODULE__{
+      document: document,
+      quoted_document: quoted_document,
+      entries: [],
```

Comment:
Will a list be able to elegantly support incremental indexing (when documents change)? We'll need to find and remove entries from the list on top of adding new ones, so I'm not quite sure a flat list is going to scale very far.

The first alternative that comes to mind is a map, though finding the right key for that might be a challenge.

I know we're not at the optimization phase yet, but since this is kind of the core of the system it'd be nice not to paint ourselves into a corner.

Reply:
I may be mistaken, but I take this to be a sort of basic "collect all the things", not the final storage mechanism.

Reply:
That's correct, @zachallaun. 
I'm currently experimenting with ETS and sqlite as storage mechanisms for the on-disk index, and yes, incremental indexing is very important.

Reply:
Ok so the reducer would only built a list of entries for a single indexing pass, and then those entries will get passed somewhere else for storage in a search-friendly format. Makes sense!

Reply:
That's correct, @Blond11516. Check out my latest commit for where it all goes.